onnxruntime/core/optimizer/computation_reduction.cc

```diff
@@ -0,0 +1,218 @@
+  // because GatherND's batch_dims be 1).
+  // otherwise, it might not be safe to move GatherND upward.
+  for (int i = 0; i < 2; i++) {
+    auto& output_dim = output_shape->dim(i);
+    auto& input_dim = input_shape->dim(i);
+    if (output_dim.has_dim_value() && input_dim.has_dim_value()) {
+      ORT_RETURN_IF_NOT(output_dim.dim_value() == input_dim.dim_value());
+    } else if (output_dim.has_dim_param() && input_dim.has_dim_param()) {
+      ORT_RETURN_IF_NOT(output_dim.dim_param() == input_dim.dim_param());
+    }
+  }
+  return SwapGatherNDWithInputNode(graph, gathernd_node, gathernd_input_node, 0);
+}
+
+static std::unordered_map<std::string, Handler> handlers = {
+    {"Add", AddHandler},
+    {"Gelu", SimpleHandler},
+    {"LayerNormalization", SimpleHandler},
+    {"MatMul", MatMulHandler},
+};
+
+static Status Delegate(std::string op_type, Graph& graph, Node* gathernd_node, Node* gathernd_input_node) {
+  if (handlers.count(op_type)) {
+    return handlers[op_type](graph, gathernd_node, gathernd_input_node);
+  } else {
+    return common::Status(common::ONNXRUNTIME, common::NOT_IMPLEMENTED, op_type + " handler is not implemented");
+  }
+}
+
+Status ComputationReductionTransformer::ApplyImpl(Graph& graph, bool& modified, int graph_level,
+                                                  const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& order = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto index : order) {
+    auto* node_ptr = graph.GetNode(index);
+    if (!node_ptr)
+      // node was removed, this should not happen since we are not removing nodes.
+      continue;
+
+    auto& node = *node_ptr;
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "GatherND", {1}, kOnnxDomain) ||
+        !graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders()) ||
+        node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    auto batch_dims = static_cast<int64_t>(node.GetAttributes().at("batch_dims").i());
+    if (batch_dims != 1) {
+      continue;
+    }
+
+    auto indices_shape = node.MutableInputDefs()[1]->Shape();
+    if (indices_shape == nullptr) {
+      continue;
+    }
+
+    const auto indices_rank = indices_shape->dim_size();
+    auto& indices_last_dim = indices_shape->dim(indices_rank - 1);
+    // Since GatherND is assumed to have batch_dims=1, if the input data's shape is [batch, sequence, ..., ... ],
+    // limiting indices_rank = 3 will make sure produced output is in shape [batch, sliced_sequence, ..., ...]
+    // and the rank did not change.
+    if (!(indices_last_dim.has_dim_value() && indices_last_dim.dim_value() == 1 && indices_rank == 3)) {
+      continue;
+    }
+
+    // Todo: check whether we want to move GatherND up, for example, if GatherND's outputs are larger
+    // than inputs, we should probably bring it ahead.
```

Comment:
Do we expect other conditions that we don't have to move? like inputs/outputs are smaller than certain threshold? Or something else ...

Reply:
Yeah, possibly, while for cases inputs/outputs are smaller than certain thread hold, doing this won't bring bad impact, right? 

An embarrassing thing is: with symbolic dims, we cannot compare the sizes . 
Any idea about this? @ytaous  @SherlockNoMad  @souptc   Otherwise, we have to NOT check that currently.